exonum-java-binding/qa-service/src/main/java/com/exonum/binding/qaservice/QaServiceImpl.java

```diff
@@ -229,6 +281,72 @@ public void applyConfiguration(Fork fork, Configuration configuration) {
     updateTimeOracle(fork, configuration);
   }
 
+  @Override
+  @Transaction(CREATE_COUNTER_TX_ID)
+  public void createCounter(TxMessageProtos.CreateCounterTxBody arguments, TransactionContext context)
+      throws TransactionExecutionException {
+    String name = arguments.getName();
+    QaSchema schema = new QaSchema(context.getFork(), context.getServiceName());
+    MapIndex<HashCode, Long> counters = schema.counters();
+    MapIndex<HashCode, String> names = schema.counterNames();
+
+    HashCode counterId = Hashing.defaultHashFunction()
+        .hashString(name, UTF_8);
+    if (counters.containsKey(counterId)) {
+      throw new TransactionExecutionException(COUNTER_ALREADY_EXISTS.code);
+    }
+    assert !names.containsKey(counterId) : "counterNames must not contain the id of " + name;
+
+    counters.put(counterId, 0L);
+    names.put(counterId, name);
+  }
+
+  @Override
+  @Transaction(INCREMENT_COUNTER_TX_ID)
+  public void incrementCounter(TxMessageProtos.IncrementCounterTxBody arguments, TransactionContext context)
+      throws TransactionExecutionException {
+    byte[] rawCounterId = arguments.getCounterId().toByteArray();
+    HashCode counterId = HashCode.fromBytes(rawCounterId);
+
+    QaSchema schema = new QaSchema(context.getFork(), context.getServiceName());
+    ProofMapIndexProxy<HashCode, Long> counters = schema.counters();
+
+    // Increment the counter if there is such.
+    if (!counters.containsKey(counterId)) {
+      throw new TransactionExecutionException(UNKNOWN_COUNTER.code);
+    }
+    long newValue = counters.get(counterId) + 1;
+    counters.put(counterId, newValue);
+  }
+
+  @Override
+  @Transaction(VALID_THROWING_TX_ID)
+  public void throwing(TxMessageProtos.ThrowingTxBody arguments, TransactionContext context)
+      throws TransactionExecutionException {
+    QaSchema schema = new QaSchema(context.getFork(), context.getServiceName());
+
+    // Attempt to clear all service indices.
+    schema.clearAll();
+
+    // Throw an exception. Framework must revert the changes made above.
+    throw new IllegalStateException("#execute of this transaction always throws: " + this);
```

Comment:
`this` is now the service, hence no longer appropriate. I think a hash of the message would be better (it identifies the tx message) + arguments.seed()